packages/plc/src/server/db.ts

```diff
@@ -0,0 +1,164 @@
+
+  constructor(db: SafeDB) {
+    this.db = db
+  }
+
+  static async sqlite(location: string): Promise<Database> {
+    const db = await SafeDB.sqlite(location)
+    return new Database(db)
+  }
+
+  static async memory(): Promise<Database> {
+    const db = await SafeDB.memory()
+    return new Database(db)
+  }
+
+  async close(): Promise<void> {
+    await this.db.close()
+  }
+
+  async validateAndAddOp(did: string, proposed: t.Operation): Promise<void> {
+    await this.db.transaction(async (tx) => {
+      const ops = await this._opsForDid(tx, did)
+      // throws if invalid
+      const { nullified } = await document.assureValidNextOp(did, ops, proposed)
+      const cid = await cidForData(proposed)
+
+      const toInsert = new OperationsTable()
+      toInsert.did = did
+      toInsert.operation = JSON.stringify(proposed)
+      toInsert.cid = cid.toString()
+      toInsert.createdAt = new Date()
+      await tx.save(toInsert)
+
+      if (nullified.length > 0) {
+        const nullfiedStrs = nullified.map((cid) => cid.toString())
+        await tx.update(
+          OperationsTable,
+          { did, cid: In(nullfiedStrs) },
+          { nullified: true },
+        )
+      }
+    })
+  }
+
+  async opsForDid(did: string): Promise<t.Operation[]> {
+    return this.db.readOperation(async (manager) => {
+      const res = await this._opsForDid(manager, did)
+      return res.map((row) => row.operation)
+    })
+  }
+
+  // helper fn that returns additional information
+  async _opsForDid(
+    manager: EntityManager,
+    did: string,
+  ): Promise<t.IndexedOperation[]> {
+    const res = await manager.findBy(OperationsTable, {
+      did,
+      nullified: false,
+    })
+    return res
+      .map((row) => ({
+        did: row.did,
+        operation: JSON.parse(row.operation),
+        cid: CID.parse(row.cid),
+        nullified: row.nullified,
+        createdAt: row.createdAt,
+      }))
+      .sort((a, b) => {
+        return a.createdAt.getTime() - b.createdAt.getTime()
+      })
+  }
+}
+
+// It is VERY IMPORTANT that any writes go through the `transaction` method on the underlying SafeDB
+// This is needed because sqlite does not offer a connection pool & transactions can end up getting nested
+// and either fail or return successful but then get rolledback if the outer tx fails
+class SafeDB {
+  db: DataSource
+  lock: Mutex
+
+  constructor(db: DataSource) {
+    this.db = db
+    this.lock = new Mutex()
+  }
+
+  static async sqlite(location: string): Promise<SafeDB> {
+    const db = new DataSource({
+      type: 'sqlite',
+      database: location,
+      entities: [OperationsTable],
+      synchronize: true,
+    })
+    await db.initialize()
+    return new SafeDB(db)
+  }
+
+  static async memory(): Promise<SafeDB> {
+    return SafeDB.sqlite(':memory:')
+  }
+
+  async close(): Promise<void> {
+    await this.db.destroy()
+  }
+
+  async transaction<T>(fn: (manager: EntityManager) => Promise<T>): Promise<T> {
```

Comment:
Okay so Sqlite only has one connection at a time & neither sqlite nor typeorm implement some sort of tx queue/pool: https://github.com/typeorm/typeorm/issues/307

We actually have two separate consistency problems here:
- the normal reason for using sql txs: on recovery operations we nullify some operations & add a new operations. these need to happen atomically
- our DB needs to be _strongly_ consistent. An new operation should not be added to the DB unless we know _for sure_ that no other operations have happened in the meantime. Otherwise we fork the DID

The first requires implementing a tx queue around writes.
The second requires a mutex on individual DIDs.

To avoid the complexity of competing mutexes, I simply implemented a mutex on txs. & do both the read and the write within that tx. 

I broke out the more costly validation step to be a simple read outside of the tx and just do a quick check within the tx that the prev value matches the expected value